### PR TITLE
[stable/datadog] Add options to use new cluster-agent

### PR DESCRIPTION
**Pull request by coreypobrien**
This adds the option to use to the new beta cluster-agent implementation (details [here](https://github.com/DataDog/datadog-agent/tree/master/docs/cluster-agent)).

Everything is off by default except that the default image is upgraded from 6.3.2 to 6.4.2.

Here's what it looks like with `clusterAgent.enabled=true` and `clusterAgent.metricsProvider.enabled=true`

```
$ helm status datadog
LAST DEPLOYED: Wed Sep  5 22:49:22 2018
NAMESPACE: kube-system
STATUS: DEPLOYED

RESOURCES:
==> v1/Secret
NAME                    TYPE    DATA  AGE
datadog-cluster-agent   Opaque  1     8m
datadog-datadog         Opaque  1     8m
datadog-datadog-appkey  Opaque  1     8m

==> v1/ServiceAccount
NAME                        SECRETS  AGE
datadog-kube-state-metrics  1        8m
datadog-cluster-agent       1        8m
datadog-datadog             1        8m

==> v1beta1/ClusterRoleBinding
NAME                        AGE
datadog-kube-state-metrics  8m

==> v1/RoleBinding
NAME                   AGE
datadog-cluster-agent  8m

==> v1beta1/APIService
NAME                             AGE
v1beta1.external.metrics.k8s.io  8m

==> v1/Pod(related)
NAME                                         READY  STATUS   RESTARTS  AGE
datadog-datadog-2n5d4                        1/1    Running  0         8m
datadog-datadog-b7cjp                        1/1    Running  0         8m
datadog-datadog-d67xv                        1/1    Running  0         8m
datadog-datadog-l6kxw                        1/1    Running  0         8m
datadog-datadog-r7lqz                        1/1    Running  0         8m
datadog-datadog-xlzzg                        1/1    Running  0         8m
datadog-kube-state-metrics-7dc475d548-x99jr  1/1    Running  0         8m
datadog-cluster-agent-64486fd9bf-2zssr       1/1    Running  0         8m

==> v1beta1/ClusterRole
NAME                        AGE
datadog-kube-state-metrics  8m

==> v1/ClusterRole
datadog-cluster-agent                          8m
datadog-datadog                                8m
datadog-cluster-agent-external-metrics-reader  8m

==> v1/ClusterRoleBinding
NAME                                           AGE
datadog-cluster-agent:system:auth-delegator    8m
datadog-cluster-agent                          8m
datadog-datadog                                8m
datadog-cluster-agent-external-metrics-reader  8m

==> v1/Service
NAME                               TYPE       CLUSTER-IP      EXTERNAL-IP  PORT(S)   AGE
datadog-kube-state-metrics         ClusterIP  100.65.247.192  <none>       8080/TCP  8m
datadog-cluster-agent              ClusterIP  100.71.176.251  <none>       5005/TCP  8m
datadog-cluster-agent-metrics-api  ClusterIP  100.65.42.35    <none>       443/TCP   8m

==> v1beta1/DaemonSet
NAME             DESIRED  CURRENT  READY  UP-TO-DATE  AVAILABLE  NODE SELECTOR  AGE
datadog-datadog  6        6        6      6           6          <none>         8m

==> v1beta1/Deployment
NAME                        DESIRED  CURRENT  UP-TO-DATE  AVAILABLE  AGE
datadog-kube-state-metrics  1        1        1           1          8m
datadog-cluster-agent       1        1        1           1          8m


NOTES:

DataDog agents are spinning up on each node in your cluster. After a few
minutes, you should see your agents starting in your event stream:

    https://app.datadoghq.com/event/stream
```

### CODE REVIEW

File: stable/datadog/Chart.yaml
Comment (davidkarlsen) [inline stable/datadog/Chart.yaml]: update this to match the upgrade

File: stable/datadog/templates/agent-deployment.yaml
Comment (davidkarlsen) [inline stable/datadog/templates/agent-deployment.yaml]: this can be v1 as latest and previous k8s api is what is standard to support: https://v1-10.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#deployment-v1-apps

**Review by davidkarlsen — CHANGES_REQUESTED** [review]

**Comment by coreypobrien** [comment]
@davidkarlsen Thanks for the feedback! Comments addressed 👍

File: stable/datadog/Chart.yaml
Comment (davidkarlsen) [reply stable/datadog/Chart.yaml]: I was thinking of appVersion - it should be the same as the images (which is the app being installed) - so 6.4.2 should be used for appVersion

Comment (davidkarlsen) [reply stable/datadog/Chart.yaml]: but 1.1.0 is also good for the chart

Comment (coreypobrien) [reply stable/datadog/Chart.yaml]: 👍 yep, I updated the appVersion in the subsequent commit

**Comment by hkaj** [comment]
Hi @coreypobrien 
Thanks for the PR!

We are very excited to see interest in the cluster agent! OTOH we don't think it's ready for prime time just yet. There are several things we want to improve before releasing it on our users' clusters. We (and by we I mean @CharlyF 😁 ) will look at your PR as soon as we have a build that satisfies our QA.
Thanks again for contributing

**Comment by coreypobrien** [comment]
@hkaj I definitely understand not making the default, which is why this PR keeps all the default behavior of not using the cluster-agent. I think having the _option_ to use it is useful since we'll be adding it into the chart eventually, right?

**Review by davidkarlsen — APPROVED** [review]

**Comment by davidkarlsen** [comment]
@hkaj @coreypobrien I'm fine with it - and also like the general upgrade.
Since default is to leave cluster-agent disabled - should we run with it?
If it is flakey - maybe some additional notes in the README to warn about maturity?

**Comment by CharlyF** [comment]
Hey @coreypobrien @davidkarlsen, I've been leading the Cluster Agent project and I am very happy to see this contrib!
I wanted to weigh in to further explain the rationale behind our wish not to merge this _just_ yet.
While it is indeed still in beta and disabled by default, we are still at a stage where we try to connect with individual users to receive feedback and make this Cluster Agent GA ready. 
Merging this would expose the Cluster Agent's option and I believe that along with any feature, beta or not, comes a set of expectations. 
As we want to deliver the best experience possible, I would really appreciate holding off on this for a short amount of time until we have collected enough feedback that we can expose this option here (disabled by default and with the appropriate documentation).
Let me know if this makes sense, and again I really appreciate seeing such a traction.
Best,
.C

**Comment by coreypobrien** [comment]
@CharlyF I think the distinction I see is between the default and non-default. In the same way that you have already published the cluster-agent images to docker hub and the code and instructions in github, adding support to the helm chart here is work to be done _before_ making it the recommended solution. In a classic chicken-egg fashion, I'd love to evaluate and give feedback, but I can't do that well without being able to reliably deploy it to my kubernetes cluster using helm.

**Comment by coreypobrien** [comment]
@CharlyF Could we go ahead and merge this now?

File: stable/datadog/templates/daemonset.yaml
Comment (CharlyF) [inline stable/datadog/templates/daemonset.yaml]: Could we use `{{ .Values.clusterAgent.enabled }}` ?

File: stable/datadog/templates/agent-clusterrole.yaml
Comment (CharlyF) [inline stable/datadog/templates/agent-clusterrole.yaml:14]: if the cluster agent is used, we could trim the clusterrole of the node agent down to:
```
- nonResourceURLs:
  - "/version"
  - "/healthz"
  verbs:
  - get
```
As the node agent only needs to communicate with the kubelet.
We would also want to add:
```
- apiGroups:
  - ""
  resources:
  - nodes
  verbs:
  - get
``` 
As the option NODE_LABELS_AS_TAGS is only supported in 6.5+. So we can add it to the list above.

As you are only using one cluster role shared by both, I'm happy to leave it as is for now, and we can make this change later.

File: stable/datadog/templates/agent-deployment.yaml
Comment (CharlyF) [inline stable/datadog/templates/agent-deployment.yaml:83]: probing the metrics provider is a great idea, we are currently rethinking how we do liveness probes, I will definitely keep this in mind.

File: stable/datadog/README.md
Comment (CharlyF) [inline stable/datadog/README.md:87]: Let's add that it only supports Kubernetes 1.10+ here.

**Review by CharlyF — CHANGES_REQUESTED** [review]
@coreypobrien Thanks again for this contribution. This is a significant milestone.
I left a few comments on this, once addressed, we can merge it.

File: stable/datadog/templates/agent-clusterrole.yaml
Comment (coreypobrien) [reply stable/datadog/templates/agent-clusterrole.yaml:14]: We're using 3 clusterroles actually. One for the node agent that is currently set to

```
- apiGroups:
  - ""
  resources:
  - nodes/metrics
  - nodes/spec
  - nodes/proxy
  verbs:
  - get
```

a more expansive one for the cluster-agent which is what this file covers, and a third role in `agent-clusterrolebinding-auth-delegator.yaml` that is used in addition to `agent-clusterrole.yaml` if the metrics server is enabled.

I based these changes on https://github.com/DataDog/datadog-agent/tree/master/Dockerfiles/manifests/cluster-agent which seem to contradict your suggestion here (e.g. the node agent doesn't have `nonResourceURLs`).

Comment (davidkarlsen) [reply stable/datadog/templates/agent-clusterrole.yaml:14]: @CharlyF Are we ok with this one now and can merge?

**Comment by CharlyF** [comment]
This looks good to me, thanks again!

**Pull request merged**
